core/src/main/java/com/google/cloud/sql/core/CloudSqlInstance.java

```diff
@@ -100,13 +113,15 @@
    * no valid data is currently available.
    */
   private InstanceData getInstanceData() {
-    try {
-      // TODO(kvg): Let exceptions up to here before adding context
-      return Uninterruptibles.getUninterruptibly(currentInstanceData);
-    } catch (ExecutionException ex) {
-      Throwable cause = ex.getCause();
-      Throwables.throwIfUnchecked(cause);
-      throw new RuntimeException(cause);
+    synchronized (instanceDataGuard) {
```

Comment:
I think if you synchronize inside the try, you'll release the lock earlier on an exception. Probably doesn't matter in practice, but generally synchronization blocks should be as small as possible.

Now that I think about it, you actually don't want to synchronize while waiting at all. Try this:

```java
ListenableFuture<InstanceData> instanceData;
synchronized (instanceDataGuard) {
  instanceData = currentInstanceData;
}
try {
 return getUninterruptibly(instanceData);
} catch (...) { ... }
```

Now you've released the lock before you block.

Reply:
In this case, do we even need to synchronize? I thought references assignments were already atomic in java, so doesn't calling the function already use the value of the reference?

Reply:
They are if you make the field `volatile`.

If you're using synchronized blocks, you really should synchronize around reads and writes. And since you need to synchronize across two fields, `volatile` isn't enough. I was hoping we could get down to one field so that you could just make it `volatile`, but we haven't.